File: campaigner/lib/seasonal.py
"""
Seasonal multiplier — manual bridge to War Chest (v2).

Per decisions-log §1.10: `businesses.seasonal_hints` holds windows the
operator enters by hand. Each window has {start, end, multiplier, confidence}.
On any given date, the effective multiplier is the PRODUCT of multipliers of
all windows that cover that date.

Empty hints → 1.0. Overlap → product (see entry §1.10 "מה נשאר פתוח" for why).

Future v2: same shape, confidence='learned' rows get priority weighting. For
now every window is user_stated and contributes equally.
"""

from __future__ import annotations

from datetime import date
from typing import Any


def _parse_date(raw: Any) -> date | None:
    if isinstance(raw, date):
        return raw
    if isinstance(raw, str):
        try:
            return date.fromisoformat(raw)
        except ValueError:
            return None
    return None


def active_windows(seasonal_hints: dict | None, on: date) -> list[dict]:
    """Return windows from `seasonal_hints` that cover `on` (inclusive start/end)."""
    if not seasonal_hints or not isinstance(seasonal_hints, dict):
        return []
    windows = seasonal_hints.get("windows") or []
    if not isinstance(windows, list):
        return []
    active: list[dict] = []
    for w in windows:
        if not isinstance(w, dict):
            continue
        start = _parse_date(w.get("start"))
        end = _parse_date(w.get("end"))
        if start is None or end is None or start > end:
            continue
        if start <= on <= end:
            active.append(w)
    return active


def multiplier_for_date(seasonal_hints: dict | None, on: date) -> float:
    """
    Product of multipliers across all active windows on `on`.

    No active windows → 1.0. Malformed / missing multiplier on a window is
    treated as 1.0 for that window (skipped rather than crashed — seasonal
    hints are operator-entered and shouldn't brick the daily run).
    """
    total = 1.0
    for w in active_windows(seasonal_hints, on):
        m = w.get("multiplier")
        if isinstance(m, int | float) and m > 0:
            total *= float(m)
    return total


def effective_monthly_budget(
    monthly_budget_ils: float | None,
    seasonal_hints: dict | None,
    on: date,
) -> tuple[float, float, list[dict]]:
    """
    Returns (effective_budget, multiplier, active_windows).

    `monthly_budget_ils` None → (0.0, multiplier, windows). Caller decides what
    an unset budget means (typically: skip pace check, log observation).
    """
    m = multiplier_for_date(seasonal_hints, on)
    base = float(monthly_budget_ils or 0)
    return base * m, m, active_windows(seasonal_hints, on)

File: web/src/lib/seasonal.ts
/**
 * Seasonal multiplier — TS port of `campaigner/lib/seasonal.py`.
 *
 * Per decisions-log §1.10: `businesses.seasonal_hints` holds operator-entered
 * windows. On any date, the effective multiplier is the PRODUCT of multipliers
 * of all windows covering that date. Empty hints → 1.0.
 *
 * Mirrors the Python contract exactly so a row written from the web side
 * matches what the agent would write tomorrow morning.
 */

export interface SeasonalWindow {
  name?: string;
  start: string;
  end: string;
  multiplier: number;
  confidence?: string;
}

type SeasonalHintsInput = { windows?: unknown } | Record<string, unknown> | null | undefined;

function parseDate(raw: unknown): Date | null {
  if (typeof raw !== "string") return null;
  // ISO YYYY-MM-DD — interpret as UTC noon to dodge DST/TZ edge cases when
  // comparing to `on`. We only care about the calendar date.
  const m = /^(\d{4})-(\d{2})-(\d{2})$/.exec(raw);
  if (!m) return null;
  return new Date(Date.UTC(Number(m[1]), Number(m[2]) - 1, Number(m[3]), 12));
}

function sameOrBefore(a: Date, b: Date): boolean {
  return a.getTime() <= b.getTime();
}

export function activeWindows(
  seasonal_hints: SeasonalHintsInput,
  on: Date,
): SeasonalWindow[] {
  if (!seasonal_hints || typeof seasonal_hints !== "object") return [];
  const windows = (seasonal_hints as { windows?: unknown }).windows;
  if (!Array.isArray(windows)) return [];
  const active: SeasonalWindow[] = [];
  for (const w of windows) {
    if (!w || typeof w !== "object") continue;
    const ww = w as Record<string, unknown>;
    const start = parseDate(ww.start);
    const end = parseDate(ww.end);
    if (!start || !end || start.getTime() > end.getTime()) continue;
    if (sameOrBefore(start, on) && sameOrBefore(on, end)) {
      active.push({
        name: typeof ww.name === "string" ? ww.name : undefined,
        start: ww.start as string,
        end: ww.end as string,
        multiplier:
          typeof ww.multiplier === "number" && ww.multiplier > 0
            ? ww.multiplier
            : 1.0,
        confidence:
          typeof ww.confidence === "string" ? ww.confidence : undefined,
      });
    }
  }
  return active;
}

export function multiplierForDate(
  seasonal_hints: SeasonalHintsInput,
  on: Date,
): number {
  let total = 1.0;
  for (const w of activeWindows(seasonal_hints, on)) {
    if (typeof w.multiplier === "number" && w.multiplier > 0) {
      total *= w.multiplier;
    }
  }
  return total;
}

export function effectiveMonthlyBudget(
  monthly_budget_ils: number | null | undefined,
  seasonal_hints: SeasonalHintsInput,
  on: Date,
): {
  effective: number;
  multiplier: number;
  active_windows: SeasonalWindow[];
} {
  const m = multiplierForDate(seasonal_hints, on);
  const base = Number(monthly_budget_ils ?? 0);
  return {
    effective: base * m,
    multiplier: m,
    active_windows: activeWindows(seasonal_hints, on),
  };
}
